File: 프로그래머스/3/60062. 외벽 점검/외벽 점검.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int solution(int n, vector<int> weak, vector<int> dist) {
    
    
    int w_size = weak.size();
    int d_size = dist.size();
    for(int i=0; i<w_size; i++) {
        weak.push_back(n + weak[i]);
    }
    
    int answer = d_size+1;
    
    sort(dist.begin(), dist.end());
    for(int s = 0; s < w_size; s++){
        do {
            int cnt = 1;
            int pos = weak[s] + dist[cnt - 1];
            
            for (int i=s; i < s+w_size; i++) {
                if (pos < weak[i]) {
                    cnt++;
                    
                    if (cnt > d_size) break;
                    
                    pos = weak[i] + dist[cnt-1];
                }    
            }
            
            if (cnt <= d_size) {
                answer = min(answer, cnt);
            }
        } while(next_permutation(dist.begin(), dist.end()));    
    }
    
    
    
    return answer > d_size ? -1 : answer;
}

File: 프로그래머스/3/60062. 외벽 점검/외벽 점검.java
class Solution {
    int n, answer;
    int[] weak, dist;
    int[] extendedWeak;
    boolean[] visited;
    
    public int solution(int n, int[] weak, int[] dist) {
        this.n = n;
        this.weak = weak;
        this.dist = dist;
        this.answer = dist.length + 1;
        this.visited = new boolean[dist.length];
        
        extendedWeak = new int[weak.length * 2];
        
        for(int i=0; i<weak.length; i++){
            extendedWeak[i] = weak[i];
            extendedWeak[i + weak.length] = weak[i] + n;
        }
        
        
        for(int i=0; i<weak.length; i++){
            dfs(i, 0, 0);
        }
        
        return answer > dist.length ? -1 : answer;
    }
    
    void dfs(int start, int count, int repairedCount) {
        
        if (count >= answer) return;
        
        if (repairedCount >= weak.length) {
            answer = Math.min(answer, count);
            return;
        }
        
        
        for (int i=0; i<dist.length; i++) {
            if (!visited[i]) {
                visited[i] = true;
                
                int currentCover = extendedWeak[start] + dist[i];
                int nextRepaired = 0;
                
                while (start + nextRepaired < extendedWeak.length &&
                      extendedWeak[start + nextRepaired] <= currentCover) {
                    nextRepaired++;
                }
                
                dfs(start + nextRepaired, count + 1, repairedCount + nextRepaired);
                
                visited[i] = false;
                
            }
        }
    }
}